File: golang-devildyw/middle/L207课程表.go
package main

func main() {}

func canFinish(numCourses int, prerequisites [][]int) bool {
	// 拓扑排序
	// 1. 入度数组 记录着节点的入度（代表着指向节点边的数量）
	// 2. 邻接表（记录着某一个节点 他的指向 可能有多个指向所以用map[int][]int 来保存）
	// 3. 队列（记录入度为0的节点）

	// 入度表
	inDegree := make([]int, numCourses)
	// 邻接表
	adj := make(map[int][]int)

	for _, edge := range prerequisites {
		cur := edge[0]
		pre := edge[1]
		inDegree[cur]++
		// 构建邻接表：pre 课指向cur课
		adj[pre] = append(adj[pre], cur)
	}

	//3. 将所有入度为0的课入列
	queue := []int{}
	for i := 0; i < len(inDegree); i++ {
		if inDegree[i] == 0 {
			queue = append(queue, i)
		}
	}

	//4. 开始选课
	count := 0
	for len(queue) > 0 {
		//出队
		selected := queue[0]
		queue = queue[1:]

		count++

		//学习了这门课之后需要减少相应的入度 这个从邻接表看
		toEnQueue := adj[selected]
		if len(toEnQueue) > 0 {
			for _, nextCourse := range toEnQueue {
				// 后续课程入度-1
				inDegree[nextCourse]--
				if inDegree[nextCourse] == 0 {
					//如果入度为0了 将课程加入队列 代表课学习
					queue = append(queue, nextCourse)
				}
			}
		}
	}
	//最后判断学习的课程是否等于所需学习的课程
	return count == numCourses
}

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L207课程表.java
package top.dyw.leetcode.Hot100;

public class L207课程表 {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        // 1. 核心优化：使用数组代替 HashMap
        // 因为课程编号是 0 ~ numCourses-1，天然适合数组下标
        List<Integer>[] adj = new ArrayList[numCourses];
        int[] inDegree = new int[numCourses];

        // 2. 预先初始化所有的 List，避免在循环中判空或创建
        for (int i = 0; i < numCourses; i++) {
            adj[i] = new ArrayList<>();
        }

        // 3. 构建图
        for (int[] p : prerequisites) {
            // 题目描述：想修课程 p[0]，先修 p[1]。
            // 依赖关系： p[1] -> p[0]
            int course = p[0];
            int pre = p[1];

            adj[pre].add(course); // 邻接表记录指向
            inDegree[course]++;   // 入度增加
        }

        // 4. 将入度为 0 的节点加入队列
        // 使用 ArrayDeque 比 LinkedList 稍微快一点点（数组实现 vs 链表实现）
        Deque<Integer> queue = new ArrayDeque<>();
        for (int i = 0; i < numCourses; i++) {
            if (inDegree[i] == 0) {
                queue.offer(i);
            }
        }

        int count = 0; // 记录已修课程数量

        // 5. BFS
        while (!queue.isEmpty()) {
            int cur = queue.poll();
            count++;

            // 直接通过数组下标拿邻居，不需要哈希查找，速度极快
            for (int neighbor : adj[cur]) {
                inDegree[neighbor]--;
                if (inDegree[neighbor] == 0) {
                    queue.offer(neighbor);
                }
            }
        }

        return count == numCourses;
    }
}
